proposals/010-authorizer.md

```diff
@@ -0,0 +1,212 @@
+| CONNECT   | Cluster       | SaslAuthenticate                                                                                                                                                                                                       |
+
+In general, the filter will make access decisions in the same manner as Kafka itself.  This means it will apply the same authorizer checks that Kafka enforces itself and generate error responses in the same way.  It will
+From the client's perspective, it will be impossible for it to distinguish between the proxy and kafka cluster itself.
+It will also use the same implied operation semantics as implemented by Kafka itself, such as where `ALTER` implies `DESCRIBE`, as described by 
+`org.apache.kafka.common.acl.AclOperation`.
+
+There is one deviation.  The filter will implement a `CONNECT` authorization check on the `CLUSTER` early, as the connection is made, once the principal is known.  This allows the Authorizer filter to be used to gate access to virtual clusters.
+* In the case of SASL, this will be performed on receipt of the Sasl Authentication Response.  If the authorization check fails, the authentication will fail with an authorization failure and the connection closed.
+* In case of TLS client-auth, this will be performed on receipt of the first request message.   If the authorization check fails, a short circuit response will be sent containing an authorization failure and the connection closed.  This feature won’t be part of the initial scope.
+
+The filter will support messages that express topic identity using topic ids (i.e. those building on [KIP-516](https://cwiki.apache.org/confluence/display/KAFKA/KIP-516%3A+Topic+Identifiers)).  It will resolve the topic id into a topic name before making the authorization check.
+
+### File based Authorizer implementation
+
+The initial scope will include an Authorizer implementation that is backed by authorization rules expressed in a separate file.  This file will associate principals with ACL rules capable of expressing an allow-list of resources.
+The initial version will be restricted to expressing allow lists of topics, but future version will extend this to allow for rules to be express about other resource types.
+
+### APIs
+
+#### Authorizer Filter
+
+Filter Configuration:
+
+```yaml
+type: AuthorizerFilter
+config:
+  authorizer: FileBasedAllowListAuthorizer
+  authorizerConfig:
+    rulesFile: /path/to/allow-list.yaml
+```
+
+Java APIs:
+
+```java
+// Inspired by org.apache.kafka.server.authorizer.Authorizer, except is asynchronous in nature.
+interface Authorizer {
+    CompletionStage<List<AuthenticationResult>> authorize(AuthorizableRequestContext context, List<Action> actions);
+}
+```
+
+```java
+// Inspired by org.apache.kafka.server.authorizer.AuthorizableRequestContext
+interface AuthorizableRequestContext {
```

Comment:
You see to consider how you handle an anonymous client. In particular what is their principal.

Reply:
Good point. I hadn't thought about that at all.  I suppose you might have two gateways, one attached to a SASL listener and one attached to listener using anonymous.  You might want to give the anonymous connections fewer privileges. 

I need to think about this one.

Two thoughts: 
- As things stand now, as Kafka doesn't use https://www.rfc-editor.org/rfc/rfc4505, there's no SASL negotiation, so there's nothing to actually call `io.kroxylicious.proxy.filter.FilterContext#clientSaslAuthenticationSuccess`.  `#clientSaslContext` would return an empty optional.
- Kafka uses User:ANONYMOUS is its ACL, but that feels weak.  Somebody could define a user called `ANOMYMOUS` and get the rights ascribed to ANONYMOUS.  How useful is such an attack vector? I'm not sure.
-